core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/ExpressionValidator.scala

```diff
@@ -0,0 +1,48 @@
+package org.broadinstitute.dsde.rawls.expressions
+
+import org.broadinstitute.dsde.rawls.RawlsExceptionWithErrorReport
+import org.broadinstitute.dsde.rawls.jobexec.MethodConfigResolver
+import org.broadinstitute.dsde.rawls.model.{AttributeString, ErrorReport, MethodConfiguration, ValidatedMethodConfiguration}
+import spray.http.StatusCodes
+
+import scala.util.Try
+
+object ExpressionValidator {
+  // inputsToParse is parameterized in order to validate in the presence or absence of the associated Method
+  // presence: inputs which are both empty and optional are pre-validated, so they are skipped here
+  // absence: validate all inputs normally
+
+  private[expressions] def validateAndParse(methodConfiguration: MethodConfiguration, inputsToParse: Map[String, AttributeString], parser: SlickExpressionParser) = {
+    val (emptyOutputs, outputsToParse) = methodConfiguration.outputs.partition { case (_, expr) => expr.value.isEmpty }
+
+    val parsed = ExpressionParser.parseMCExpressions(inputsToParse, outputsToParse, parser)
+
+    // empty output expressions are also valid
+    val validatedOutputs = emptyOutputs.keys.toSeq ++ parsed.validOutputs
+
+    ValidatedMethodConfiguration(methodConfiguration, parsed.validInputs, parsed.invalidInputs, validatedOutputs, parsed.invalidOutputs)
+  }
+
+  // validate a MC without retrieving the Method from the Repo: assume all inputs are required
+  def validateAndParseMCExpressions(methodConfiguration: MethodConfiguration, parser: SlickExpressionParser): ValidatedMethodConfiguration =
```

Comment:
Maybe I don't know all the context here - but why are there 2 versions, one that throws an exception and one that doesn't? I understand why we wouldn't want to launch submissions for invalid expressions -- what's the use case for _not_ throwing an exception for invalid expressions?

Reply:
In the general case, it is OK for a Method Config to exist with invalid expressions.  (I didn't realize that to be true until recently, and that false assumption is what drove GAWB-2444.)

It's about saving partial work, essentially.  If the MC has 40 inputs, we want the user to be able to fill in a few and come back later.  This no-fail validation will still inform them about what's missing but it won't block them.  If they later try to run something using the unfinished MC, then it will call the other validation and fail.

Reply:
I see, makes sense. Thanks